bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/ChannelPool.java

```diff
@@ -33,7 +35,35 @@
 
   protected static final Logger log = Logger.getLogger(ChannelPool.class.getName());
 
-  private final Channel[] channels;
+  public class WrappedChannel extends Channel {
+
+    final Channel delegate;
+    private boolean returned = false;
+
+    private WrappedChannel(Channel delegate, boolean returned) {
+      this.delegate = delegate;
+      this.returned = returned;
+    }
+
+    @Override
+    public <RequestT, ResponseT> ClientCall<RequestT, ResponseT> newCall(
+        MethodDescriptor<RequestT, ResponseT> methodDescriptor, CallOptions callOptions) {
+      return createWrappedCall(methodDescriptor, callOptions, delegate);
+    }
+
+    @Override
+    public String authority() {
+      return delegate.authority();
+    }
+
+    public synchronized void returnToPool() {
```

Comment:
I synchronize returnToPool() so that checking and setting WrappedChannel.returned is atomic.  Channel.returnChannel is also synchronized.